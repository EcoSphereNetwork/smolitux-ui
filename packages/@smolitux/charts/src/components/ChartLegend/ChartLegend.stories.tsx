import React from 'react';
import { Meta, StoryObj } from '@storybook/react';
import { ChartLegend } from './ChartLegend';

const meta: Meta<typeof ChartLegend> = {
  title: 'Charts/ChartLegend',
  component: ChartLegend,
  parameters: { layout: 'centered' },
  tags: ['autodocs'],
  argTypes: {
    direction: { control: { type: 'radio' }, options: ['horizontal', 'vertical'] },
  },
};
export default meta;

type Story = StoryObj<typeof ChartLegend>;

const legendItems = [
  { label: 'Series A', color: '#ff0000' },
  { label: 'Series B', color: '#00ff00' },
];

export const Vertical: Story = {
  render: (args) => (
    <svg width={200} height={80} viewBox="0 0 200 80">
      <ChartLegend {...args} />
    </svg>
  ),
  args: {
    items: legendItems,
    direction: 'vertical',
  },
};

export const Horizontal: Story = {
  render: (args) => (
    <svg width={200} height={40} viewBox="0 0 200 40">
      <ChartLegend {...args} />
    </svg>
  ),
  args: {
    items: legendItems,
    direction: 'horizontal',
  },
};
