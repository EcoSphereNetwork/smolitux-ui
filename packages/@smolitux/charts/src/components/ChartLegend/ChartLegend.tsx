import React, { forwardRef } from 'react';

export interface LegendItem {
  /** Beschriftung des Eintrags */
  label: string;
  /** Farbe des Eintrags */
  color: string;
}

export interface ChartLegendProps extends React.SVGAttributes<SVGGElement> {
  /** Einträge der Legende */
  items: LegendItem[];
  /** Ausrichtung der Einträge */
  direction?: 'horizontal' | 'vertical';
  /** Abstand zwischen den Einträgen */
  itemSpacing?: number;
  /** Größe der Farbmarkierung */
  markerSize?: number;
}

/**
 * Wiederverwendbare SVG-Legende
 */
export const ChartLegend = forwardRef<SVGGElement, ChartLegendProps>(
  (
    {
      items,
      direction = 'vertical',
      itemSpacing = 20,
      markerSize = 12,
      className = '',
      ...rest
    },
    ref,
  ) => {
    return (
      <g ref={ref} className={`smolitux-chart-legend ${className}`.trim()} {...rest} data-testid="chart-legend">
        {items.map((item, index) => {
          const translate =
            direction === 'vertical'
              ? `translate(0, ${index * itemSpacing})`
              : `translate(${index * (markerSize * 4)}, 0)`;
          return (
            <g key={item.label} transform={translate} className="legend-item">
              <rect width={markerSize} height={markerSize} fill={item.color} rx={2} />
              <text x={markerSize + 4} y={markerSize / 2} dominantBaseline="middle" fontSize="12" fill="currentColor">
                {item.label}
              </text>
            </g>
          );
        })}
      </g>
    );
  },
);

ChartLegend.displayName = 'ChartLegend';

export default ChartLegend;
