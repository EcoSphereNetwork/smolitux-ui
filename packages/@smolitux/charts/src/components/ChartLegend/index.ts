export { default, type ChartLegendProps, type LegendItem } from './ChartLegend';
