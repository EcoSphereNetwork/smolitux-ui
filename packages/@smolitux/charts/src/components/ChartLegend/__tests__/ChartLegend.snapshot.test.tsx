import React from 'react';
import { render } from '@testing-library/react';
import { ChartLegend } from '../ChartLegend';

describe('ChartLegend Snapshot', () => {
  it('matches snapshot', () => {
    const items = [{ label: 'A', color: '#ff0000' }];
    const { asFragment } = render(
      <svg>
        <ChartLegend items={items} />
      </svg>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
