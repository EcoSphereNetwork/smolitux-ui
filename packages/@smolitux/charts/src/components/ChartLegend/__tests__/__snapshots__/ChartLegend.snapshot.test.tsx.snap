// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartLegend Snapshot matches snapshot 1`] = `
<DocumentFragment>
  <svg>
    <g
      class="smolitux-chart-legend"
      data-testid="chart-legend"
    >
      <g
        class="legend-item"
        transform="translate(0, 0)"
      >
        <rect
          fill="#ff0000"
          height="12"
          rx="2"
          width="12"
        />
        <text
          dominant-baseline="middle"
          fill="currentColor"
          font-size="12"
          x="16"
          y="6"
        >
          A
        </text>
      </g>
    </g>
  </svg>
</DocumentFragment>
`;
